From 953b2adc54981efdcb3d4ac0103c626427eb1068 Mon Sep 17 00:00:00 2001
From: Kazunori Kobayashi <kkobayas@igel.co.jp>
Date: Tue, 2 Oct 2012 14:40:56 +0900
Subject: [PATCH 69/97] ext/dfbvideosink: use next_field_offset for interlace
 to progressive conversion

This patch gets next_field_offset from caps and accordingly set the
offset to the second field of Y and UV plane when receiving a field
structured image.
---
 ext/directfb/dfbvideosink.c |   11 +++++++++--
 ext/directfb/dfbvideosink.h |    1 +
 2 files changed, 10 insertions(+), 2 deletions(-)

diff --git a/ext/directfb/dfbvideosink.c b/ext/directfb/dfbvideosink.c
index 56f8946..7c28744 100644
--- a/ext/directfb/dfbvideosink.c
+++ b/ext/directfb/dfbvideosink.c
@@ -1604,6 +1604,12 @@ gst_dfbvideosink_setcaps (GstBaseSink * bsink, GstCaps * caps)
     dfbvideosink->interlaced = FALSE;
   }
 
+  if (!gst_structure_get_int (structure, "next_field_offset",
+          &dfbvideosink->next_field_offset) && dfbvideosink->interlaced) {
+    GST_ERROR_OBJECT (dfbvideosink,
+        "failed to get next-field-offset from caps in despite of interlaced being set");
+    goto beach;
+  }
 #if defined(HAVE_SHMERAM)
   stride =
       (dfbvideosink->rowstride >
@@ -2318,8 +2324,8 @@ gst_dfbvideosink_show_frame (GstBaseSink * bsink, GstBuffer * buf)
       /* Then, prepare for rendering the bottom field */
       src.h += is_src_odd;
       result.h += is_dst_odd;
-      src_datay += dfbvideosink->chroma_byte_offset / 2;
-      src_datac = src_datay + dfbvideosink->chroma_byte_offset * 3 / 4;
+      src_datay += dfbvideosink->next_field_offset;
+      src_datac += dfbvideosink->next_field_offset / 2;
       data += dest_pitch;       /* step into the next line */
       dest_pitch *= 2;          /* skip 1 line per rendering */
     } else {
@@ -3067,6 +3073,7 @@ gst_dfbvideosink_init (GstDfbVideoSink * dfbvideosink)
   dfbvideosink->rowstride = -1;
   dfbvideosink->chroma_byte_offset = -1;
   dfbvideosink->interlaced = FALSE;
+  dfbvideosink->next_field_offset = -1;
 #endif
 
   dfbvideosink->dfb = NULL;
diff --git a/ext/directfb/dfbvideosink.h b/ext/directfb/dfbvideosink.h
index 2311593..a91744e 100644
--- a/ext/directfb/dfbvideosink.h
+++ b/ext/directfb/dfbvideosink.h
@@ -136,6 +136,7 @@ struct _GstDfbVideoSink {
   gint rowstride;
   gint chroma_byte_offset;
   gboolean interlaced;
+  gint next_field_offset;
 #if defined(HAVE_SHMERAM)
   MERAM *meram;
   ICB *icby[2];
-- 
1.7.10.4

